format/spec.md

```diff
@@ -193,10 +193,38 @@ Notes:
 
 For details on how to serialize a schema to JSON, see Appendix C.
 
+#### Default value
+Default values can be assigned to top-level columns or nested fields. Default values are used during schema evolution when adding a new column. The default value is used to read rows belonging to the files that lack the column or nested field prior to the schema evolution.
```

Comment:
I think we have to hit a few points very clearly here.

1. Default values can be assigned to any newly created column in an Iceberg Table
2. Default values are only used for files which are missing the new column (Do we explicitly require the files to be written prior to the default value being set? If I import a new file directly into the iceberg table does it get the default? I think it's probably easiest to say yes)
3. New writes to the table do not use Default Values
   a. Required Columns are still required at write time
   b. Optional Columns are still persisted as null if they are missing at write time
4. Changing a default value is basically an undefined behavior
   a. Rows which have been rewritten will permanently keep the default value which was set when the row is rewritten
   b. Rows which have not been rewritten will adjust to the new default value
   c. Because of this we lock changing defaults behind "allowIncompatible changes"
5. Name Mappings are always used in place of defaults. If a file is added with Iceberg Column Ids (via import or what not) we will always use name mapping before attempting to use a column default value.
   
 I think the current text is good but can probably be tightened up.
 
 For example
 
 "Default values can be assigned to any new top-level column added to an Iceberg table." instead of the first two sentences
 "Default Values are only used when reading a data file which is missing the column with the default.

Reply:
> * New writes to the table do not use Default Values
>   a. Required Columns are still required at write time
>   b. Optional Columns are still persisted as null if they are missing at write time

I would skip "New" here. I think default values are not used on the write path at all.

Reply:
1. Agreed
2. Yes, probably use the default for imported files. But we should state that fields must be written even if there is a default value.
3. I would state this slightly differently:
    * All columns must be written in new data files at write time
    * Optional columns may contain null, required columns must not
    * Optional columns with no default value have an implied default value, `null`
    * Writers are allowed to write the default value if there is no data for a column
4. Changing the default value should not be allowed by the spec, similar to how you cannot add required columns to the spec currently. I don't think that documenting "allow incompatible changes" is a good idea -- that's just an implementation detail if you know what you're doing. Also, we will need to clarify that you may now add an optional column, an optional column with a default value, or a required column with a default value. But you cannot add a required column with no default value.
5. Agreed, but I would phrase it differently again. Name mappings are used to apply IDs for data files. Default values are added after that point, when the data file has its IDs.

Reply:
This just occurred to me and I want to note it somewhere: right now we assume that missing columns are null in a lot of places. We'll need to be careful about the implementation and make sure we catch things like Parquet file filters that assume a missing column is null.

Reply:
> * Writers are allowed to write the default value if there is no data for a column

@rdblue could you clarify what you meant by that?

Reply:
Say I have a table with schema `(id bigint NOT NULL, data string DEFAULT "--")`. Then I can write to that table and omit the `data` column:

```
INSERT INTO table VALUES (1), (2), (3);
```

The writer must actually insert `(1, "--"), (2, "--"), (3, "--)`.

This is the row with `id=3` in the behavior test:

```sql
create table default_test (id int);
insert into default_test values (1), (2);
alter table default_test add column data int default 0;
alter table default_test alter column data set default 1000;
insert into default_test (id) values (3);
insert into default_test values (4, null);
alter table default_test alter column data set default 2000;
```
```sql
select * from default_test;
```
```
id | data
-- | --
 1 | 0
 2 | 0
 3 | 1000
 4 | (null)
```

Reply:
> I would skip "New" here. I think default values are not used on the write path at all.

New is doing some useful work. Streaming engines tend to plan writes once and keep running them. And it's okay to do that. If Flink is writing with schema id=2 that is missing a column that was added in schema id=3, it can continue doing that.

Reply:
@wmoustafa and @RussellSpitzer, thinking about this a bit more, I think we may want to have a second default in the format (but not exposed to users). The first default is immutable and is set when a field is created. That tells Iceberg how to fill in the default value for all files that don't contain it (`id=1` above). The second default would be the one that is used when writing if the column isn't supplied (`id=3` above). The default for existing data files never needs to change, but the default for future writes can change at any time.

What do you think? This only needs to be visible at the format layer. Users would set a default when creating a column and could change it later to affect writes.

Reply:
@rdblue for the Writer issue. I feel like this is a little ambiguous, are writers required to use the default value with optional columns or can different writers behave differently? For me this is a little odd because we are changing the underlying table schema so I feel like the behavior shouldn't be implementation dependent.

I think we should say that using the default value for unspecified columns is the spec behavior, or say that writers should not use the value for any new values and just use null. I think giving writers the option to choose between these two behaviors is a bit tricky.

Reply:
This situation emerges when engines want to support `INSERT INTO` with a subset of columns. Note that the columns must already exist. Now, the `schema evolution` default value may or may not exist (depending on whether the column missing in `INSERT INTO` was added after the table was created (i.e., schema evolution) or existed since the table was first created (i.e., no schema evolution), respectively).

If the `schema evolution` default value exists: I think it is fair to  _reuse the value_ to fill in the missing column values in `INSERT INTO`.

If the `schema evolution` default value does not exists: I think it is fair to  _reuse the current place in the metadata_ to define this default value (upon table creation this time, and not upon schema evolution). This will never conflict with the `schema evolution` default value of the same column because the column _already exists_.

Going by the above, I think it is fine to use the same concept for the default value to serve both use cases.

For the current spec, we may state that we do not support that anyways, and hence default value cannot be leveraged on the write path. It can be a future extension though.

Reply:
Regarding the change: I believe that both can change. The only reason we cannot support change of `schema evolution` default value right now is that it is not materialized right away. Imagine we issue a complete rewrite upon `ALTER TABLE ADD col  DEFAULT 1`. The default value can be changed without any consequences (even considering `1` was the `schema evolution` one). The fact that we do not issue a full rewrite upon the DDL is an implementation detail in my opinion.

Reply:
> The default value can be changed without any consequences (even considering 1 was the schema evolution one)

@wmoustafa, this is why changing the schema evolution default is allowed as an incompatible change through the API. Mistakes happen and we can allow correcting them (for example, a typo of `DEFAULT 12` instead of `DEFAULT 1`). We just can't allow the default from when the column is added to change through a `SET DEFAULT` command.

Reply:
Sorry I should have been clearer. When I said "The default value can be changed without any consequences", I meant in an ideal situation, when there is an immediate rewrite after the `ALTER TABLE t ADD col DEFAULT 1` which will materialize the `1` in the existing rows (similar to how DBMSs do). Assuming we are in this world (`1` is materialized immediately), if later `ALTER TABLE t ALTER col SET DEFAULT 3` takes place, then this can only affect future rows when `INSERT INTO` does not specify that column (since for past rows, `1` is already in the file).  So yes, I think we are saying the same thing which `SET DEFAULT` does not change the values of existing rows, but I was saying it would be the natural behavior even if we combine both concepts to one but assuming a rewrite takes place immediately.

The main problem that I see here is that at the Iceberg API level, I expect both concepts of default values have to be exposed, at least because sometimes we say we can change it freely (for future rows) and sometimes we say we can only change it behind an `allowIncompatibleChanges` API (for existing rows). If we expose two ways or two concepts of default values, it might be unnecessarily complex.

So if I were to choose between options, both of these sounds clean:
1- Default value for `schema evolution` is the same as the one for `INSERT INTO`. None of them can change unless `allowIncompatibleChanges()` is called.
2- Default value for `schema evolution` is the same as the one for `INSERT INTO`. `schema evolution` instance is materialized right away through full rewrite upon the DDL, allowing all other instances (i.e., `INSERT INTO`) to change. From the user's point of view Default value (regardless of which instance it is) just changes without problems.

In both cases, we do not support the `INSERT INTO` use case yet.

Reply:
Replying to @RussellSpitzer:

> For me this is a little odd because we are changing the underlying table schema so I feel like the behavior shouldn't be implementation dependent.

100% agree that behavior cannot be implementation dependent!

> I feel like this is a little ambiguous, are writers required to use the default value with optional columns or can different writers behave differently?

I think that optional vs required should be a separate concept and not related to default values. Optional allows null values in the column, required means that null cannot be in the column. Then I think we have an initial default, which is the default value set when the column is added that is applied to all existing rows. Then there is also a writer default, which can change and is the value that a writer _must_ write for the column if it is not supplied by the user. This write default can change because it affects future writes and not existing data.

The place where optional and default values overlap is that we currently use `null` as the default value for optional fields. That's why I would say that optional fields have an implicit initial default of `null`. To handle this, I think we would assume that the initial default value for an existing optional field is `null` and only allow changing the write default value. When you create an optional field you can set the initial default.

Here are some more concrete evolution rules with defaults:
* When creating a required field, an initial default value must be set. Neither field values nor the default may be `null`.
* When creating an optional field, an initial default value can be set. If the initial default value is not set, it is `null`.
* The initial default value may only be set when adding a field (or through an incompatible change in the API)
* The write default value for a field starts as the initial default value, and is considered set if the field has an initial default
* The write default value may be changed
* When writing, the write default must be written into data files if a value for the column is not supplied
* When writing a field with no write default, the column must be supplied or the write must fail

Does that make it clear?

Reply:
> I was saying it would be the natural behavior even if we combine both concepts to one but assuming a rewrite takes place immediately.

This is basically what I did in the SQL snippet above, so I don't think that you can distinguish between setting the initial default and then setting the write default vs setting the initial default and then immediately changing it. But yeah, I think we're in agreement mostly.

> I expect both concepts have to be exposed somehow, at least because sometimes we say we can change it freely (for future rows) and sometimes we say we can only change it behind an allowIncompatibleChanges API (for existing rows). If we expose two ways or two concepts of default values, it might be unnecessarily complex.

In that case, I'd skip allowing users to change the initial default. If you've really just created a new column and need to update the default, you can always drop it and re-create it with the right default!

As long as we don't have that, then we can hide everything behind the API. The initial default is set when creating a field, and this also sets the write default. Setting the default value for a field only changes the write default.

Reply:
> In both cases, we do not support the INSERT INTO use case yet.

I wouldn't be surprised if we started supporting this in INSERT INTO soon. It seems like a good next step.

Reply:
> In that case, I'd skip allowing users to change the initial default. If you've really just created a new column and need to update the default, you can always drop it and re-create it with the right default!
> 
> As long as we don't have that, then we can hide everything behind the API. The initial default is set when creating a field, and this also sets the write default. Setting the default value for a field only changes the write default.

That is very close to option 1. The only difference is that option 1 allowed for changes behind the `allowIncompatibleChanges` API. Did you mean to remove that fallback? (maybe you meant we can keep it but the above is how the spec should read, which I agree with).

Reply:
It isn't close to option 1 because the write default can still freely change. The only thing we need to put behind `alloweIncompatibleChanges` is modifying the initial default.

This is fine:

```java
table.updateSchema()
    .addRequiredColumn("test", IntegerType.get(), 34) // sets the initial and write defaults to 34
    .setColumnDefault("test", 35) // sets the write default to 35
    .commit();
```

This is what would expose the initial default and would be a bit confusing:

```java
table.updateSchema()
    .addRequiredColumn("test", IntegerType.get(), 34) // sets the initial and write defaults to 34
    .commit();
table.updateSchema()
    .allowIncompatibleChanges()
    .setInitialColumnDefault("test", 35) // sets the initial default to 35
    .commit();
```

I think we could get away without adding `setInitialColumnDefault` because you can always go drop the column and re-add it if it really is a safe change. This is equivalent to the above:

```java
table.updateSchema()
    .addRequiredColumn("test", IntegerType.get(), 34) // sets the initial and write defaults to 34
    .commit();
table.updateSchema()
    .dropColumn("test")
    .commit();
table.updateSchema()
    .addRequiredColumn("test", IntegerType.get(), 35) // sets the initial and write defaults to 35
    .commit();
```

Reply:
I see. So the concern is exposing both `setColumnDefault()` and `setInitialColumnDefault()` only because we do not issue a complete rewrite upon DDL. Otherwise, we could use just one.

Reply:
Ok now that I see the code snippets, the idea is much clearer. I will refer to code snippets as (1), (2), (3), respectively.

Code snippet (1) is really nice. I think it is closer to what I was expressing in option (2) in my last comment. That is:
* It introduces the column and its schema evolution default value atomically. That would be equivalent to a rewrite, except we are storing this value in a separate placeholder instead of physically storing it with each record (i.e., rewrite).
* It reserves all future calls of `setColumnDefault()` to the future records (i.e., write side default value), and it does not override the value dedicated for schema evolution. `setColumnDefault()` can be used to change the write side default value without affecting the schema evolution default value.

I also agree (2) is confusing and it is the reason I initially thought two values can lead to such APIs.

For (3), I think it is still subject to the rules in (1), in addition to the general rules around dropping and re-introducing a column. Another way to simulate (3)'s effect without dropping and recreating the column is to issue an `UPDATE` statement from `34` to `35`. I think this has the side effect of not distinguishing between organic `34` and `34` from schema evolution, but that is how an RDBMS would behave anyways. In fact, dropping and recreating the column, along with changing default values opens many spec questions which might be beyond the scope of this PR. (For example, if you drop an existing column do you expect to wipe out the all the values (including ones on rows with non-default values) and start over if you re-introduce the column? If not what if a rewrite happens in between? Which rows does the new default value override? etc.) Also, I think this would get even more complex if we attempt to apply it on inner fields in order to update the default value of an inner field. 

So overall, I think what we have in (1) looks good. To update the past rows default value, the user can move within whatever room the rest of the spec allows them to (dropping the column if they can bear the general column dropping consequences), or running an `UPDATE` if they are okay with updating rows coincidentally carrying the default value (but RDBMSs would do the same). Finally, if really needed I think it is okay to introduce (2) to avoid jumping multiple hoops, but it should not be part of the spec. That can be done in the future as well, so we do not preemptively loosen the spec before users ask for it.

Reply:
Code snippet (3) is an alternative to code snippet (2) that doesn't expose `setInitialColumnDefault` and is demonstrating the same behavior as code snippet (1). I'm not sure what you mean that this "has the side effect of not distinguishing between organic 34 and 34 from schema evolution". This is how you would change the initial default for a column: by dropping and re-creating the column with a different default. That works because the use case is that you've made a mistake and need to correct it _before writing any data_. The point here is that we don't necessarily need to add `setInitialColumnDefault`.

The behavior of `dropColumn` is the standard behavior: the column is no longer present in the table. Rewrites are not required to keep column values for dropped columns.

Reply:
> I'm not sure what you mean that this "has the side effect of not distinguishing between organic 34 and 34 from schema evolution"

That was a caveat if the user uses `UPDATE` statement to update he default value (which is an alternative way if dropping and recreating the column is not feasible anymore due to writing more values). So after using code snippet (1), and inserting more data, users can also change the default value by issuing an UPDATE DML, but it will not only update the rows that were present before schema evolution, but also other rows after schema evolution that happened to store the same value as the default value.

> That works because the use case is that you've made a mistake and need to correct it _before writing any data_.

Agreed. It was not initially clear if the suggestion also applied to the case when more data could have been written. I guess this is where the `UPDATE` route above can be helpful.

Reply:
Sounds like we agree! I think you're right that you can use `UPDATE` to make changes as well.

Reply:
> Here are some more concrete evolution rules with defaults:
> 
> * When creating a required field, an initial default value must be set. Neither field values nor the default may be `null`.
> * When creating an optional field, an initial default value can be set. If the initial default value is not set, it is `null`.
> * The initial default value may only be set when adding a field (or through an incompatible change in the API)
> * The write default value for a field starts as the initial default value, and is considered set if the field has an initial default
> * The write default value may be changed
> * When writing, the write default must be written into data files if a value for the column is not supplied
> * When writing a field with no write default, the column must be supplied or the write must fail
> 
> Does that make it clear?

@rdblue 
I think I just want to make a clarification question so that I can make sure I understand the above discussions correctly.

So I assume to summarize this, you mean there should be 2 underlying `default value` concepts: one is `initial default value` and one is `write default value`, the `initial default value` affects old rows already written without the column and the `write default value` affects future rows.
And we only expose `one default value APIs` to users, but that API actually handles the underlying `two default value concepts`, it's just hidden to users.

Say I have a following scenario, there is a dataset with some rows written, call them `R`, a user later adds a new column with default value `d1`, and then changes the default value to `d2` (I assume that will be the `write default` value?) , is it true that according to this spec, when the user reads the dataset at this time, the `R` rows **must** read `d1` for that column, regardless of whether there is an actual full rewriting happening between the change from `d1` to `d2`? Because I think implementation-wise, we do want to lazily do the rewriting, not eagerly rewrite we set a default value?

Is my above understanding correct?

Reply:
Yes, the API only exposes one "default value". But we need to track two different values internally. One for existing data files (that cannot change) and one for future writes (than can be updated).

Reply:
> This supports lazy rewriting.

Nicely put 👍